### PR TITLE
RELION 3.0 and 3.1 STAR file support

### LINKED ISSUE #735: `RelionSource`: input starfiles from RELION `<3.1` vs `>=3.1`
Starting with RELION 3.1, a new standard was introduced for star files representing coordinates:

From https://relion.readthedocs.io/en/release-3.1/Whats-new.html

> To allow for multiple subsets of a data set having different Zernike coefficients, relion-3.1 implements the new concept of optics groups. Optics groups are defined in a separate table called data_optics at the top of a STAR file, which will also contain a table called data_movies, data_micrographs or data_particles, depending on what type of images it refers to. The second table is similar to the content of STAR files in previous releases, but contains a new column called rlnOpticsGroup, which is also present in the data_optics table. Common CTF-parameters, like rlnVoltage and _`rlnSphericalAberration, but also the new rlnOddZernike and rlnEvenZernike, can be stored once for each optics group in the data_optics table, without the need to store them for each particle/micrograph in the second table.

The code which transforms RELION starfiles into ASPIRE metadata in `RelionSource` is only set up to accept RELION 3.0 star files. RELION does have a way of transforming old style files into new ones. We should look into how to accept both types. 

Note that many datasets we work with are saved with the 3.0 format.

There is some code in `CoordinateSource` that deals with optics groups that may be useful. Also, this will allow us to read the pixel size from the STAR file, as `_rlnImagePixelSize` is a parameter also introduced with 3.1. See discussion in #322 for details on both.

### PR DESCRIPTION
**Pull request by chris-langfield**
Close #735 by supporting >= 3.1 STAR files. Also related to #309.

See section 7.7.2 in https://relion.readthedocs.io/_/downloads/en/latest/pdf/

### CODE REVIEW

**Comment by chris-langfield** [comment]
This is going to be more or less rewritten this week. I think we need a separate set of classes to specifically hold and manipulate Relion star file information. Each `Source` is trying to do different but rhyming operations, leading to confusing code.

**Comment by chris-langfield** [comment]
Summary of changes:

- Adds `RelionDataStarFile` (>= 3.1) and `RelionLegacyDataStarFile` (3.0) subclasses of `StarFile`. The `RelionDataStarFile` provides a method to "flatten" the optics group information into a single DataFrame compatible with ASPIRE metadata

- Adds a fail-first method to check off what is required to identify a valid 3.1 or 3.0 star file

- `RelionSource` now checks which version it's being given and branches accordingly.

- `CoordinateSource.import_ctf()` broken up into `import_aspire_ctf()` and `import_relion_ctf()`. The former is for a *list* of STAR files, the latter for a *single* star file (3.1 or 3.0). These both feed into `_extract_ctf()`

- `CoordinateSource` gets some minor quality of life changes and the `mrc_index` and `mrc_filepath` metadata parameters are cleaned up and set properly.

### Testing
- There is now a 3.1 and a 3.0 test Relion particles STAR file. Where available, the 3.1 (current) format STAR file has been substituted where the old one was used before.
- Test that we can instantiate `RelionSource` from either format
- Test that we can import CTFs from either format
- Validating the `apply_optics_block` method of `RelionDataStarFile`

File: tests/test_orient_sync.py [hunk 1]
```diff
@@ -95,7 +95,7 @@ def testCommandLine(self):
         with tempfile.TemporaryDirectory() as tmpdir:
             # Save the simulation object into STAR and MRCS files
             starfile_out = os.path.join(tmpdir, "save_test.star")
-            starfile_in = os.path.join(DATA_DIR, "sample_relion_data.star")
+            starfile_in = os.path.join(DATA_DIR, "sample_particles_relion31.star")
```
Comment (garrettwrong) [inline tests/test_orient_sync.py:111]: Just want to double check with you that you didn't want the 3.0 file here.

File: tests/test_coordinate_source.py [hunk 2]
```diff
@@ -316,6 +317,14 @@ def createTestRelionCtfFile(self):
         star.write(star_fp)
         return star_fp
 
+    def createTestRelionLegacyCtfFile(self):
+        legacy_star_fp = os.path.join("micrographs_ctf_legacy.star")
```
Comment (garrettwrong) [inline tests/test_coordinate_source.py]: what are you joining here?

File: tests/test_coordinate_source.py [hunk 3]
```diff
@@ -316,6 +317,14 @@ def createTestRelionCtfFile(self):
...
+        star = RelionDataStarFile(self.relion_ctf_file)
+        df = star.apply_optics_block()
+        legacy_star = StarFile(blocks=OrderedDict({"": df}))
```
Comment (garrettwrong) [inline tests/test_coordinate_source.py:328]: A few comments describing the intent of the steps here would be helpful.  I was able to work it out from your choice of var names, which is good, but it will be easier on other developers if we explain what we are doing.

File: src/aspire/utils/__init__.py [hunk 4]
```diff
@@ -62,5 +62,6 @@
     virtual_core_cpu_suggestion,
 )
 from .random import Random, choice, rand, randi, randn, random
+from .relion_interop import RelionDataStarFile, RelionLegacyDataStarFile
```
Comment (garrettwrong) [inline src/aspire/utils/__init__.py]: How do you feel about `Relion30StarFile`  and `Relion31StarFile`?

One day our current RelionDataStarFile will also be Legacy...

File: src/aspire/source/relion.py [hunk 5]
```diff
@@ -180,50 +146,59 @@ def __init__(
...
-        column_types = {
-            name: RelionSource.relion_metadata_fields.get(name, str)
-            for name in df.columns
-        }
-        df = df.astype(column_types)
+            self.data_folder = os.path.dirname(self.filepath)
+
+        rln_starfile_version = getRelionStarFileVersion(self.filepath)
+
+        if not rln_starfile_version:
```
Comment (garrettwrong) [inline src/aspire/source/relion.py]: Lets check this branch in a test since this is one of our main paths.

File: src/aspire/source/relion.py [hunk 6]
```diff
@@ -180,50 +146,59 @@ def __init__(
...
+            # get flattened data from starfile
+            metadata = starfile.apply_optics_block()
 
         # particle locations are stored as e.g. '000001@first_micrograph.mrcs'
         # in the _rlnImageName column. here, we're splitting this information
         # so we can get the particle's index in the .mrcs stack as an int
-        df[["__mrc_index", "__mrc_filename"]] = df["_rlnImageName"].str.split(
-            "@", 1, expand=True
-        )
+        metadata[["__mrc_index", "__mrc_filename"]] = metadata[
```
Comment (garrettwrong) [inline src/aspire/source/relion.py:167]: This is a much better name, thank you.

File: src/aspire/utils/relion_interop.py [hunk 7]
```diff
@@ -1,11 +1,92 @@
...
+    "_rlnMaxValueProbDistribution": float,
+    "_rlnOpticsGroup": int,
+    "_rlnOpticsGroupName": str,
+}
+
+
+def df_to_relion_types(df):
+    # convert STAR file strings to data type for each field
+    # columns without a specified data type are read as dtype=object
+    column_types = {name: relion_metadata_fields.get(name, str) for name in df.columns}
```
Comment (garrettwrong) [inline src/aspire/utils/relion_interop.py:61]: what is the relationship between dtype=object and `str`? The `get` call defaults to str. Checking that is intentional.

I just want to be careful if we are interacting with np here because

`np.array([str])` is an array of `object_`s

while an array of actual strings is different `np.array(['abc'])` is a `str_`s.

I would think the cases for arrays of floats, str, and object are distinct.  I recall there was an a bug relating to float32/64 that was potentially addressed with the use of objects. Can you point that out and how it relates to this code?

File: src/aspire/source/coordinates.py [hunk 8]
```diff
@@ -273,170 +300,119 @@ def _get_mrc_shapes(self):
...
-        self._unique_filters.append(
+
+        # get unique ctfs from the data block
+        # i'th entry of `indices` contains the index of `filter_params` with corresponding CTF params
+        filter_params, indices = np.unique(
+            df[CTF_params].astype(self.dtype).values, return_inverse=True, axis=0
+        )
+
+        # convert defocus_ang from degrees to radians
+        filter_params[:, 3] *= np.pi / 180.0
```
Comment (garrettwrong) [inline src/aspire/source/coordinates.py:371]: was this done somewhere else before, or just missing?

File: src/aspire/source/coordinates.py [hunk 9]
```diff
@@ -273,170 +300,119 @@ def _get_mrc_shapes(self):
...
+        self.unique_filters = [
             CTFFilter(
-                pixel_size=params["pixel_size"],
-                voltage=params["voltage"],
-                defocus_u=params["defocus_u"],
-                defocus_v=params["defocus_v"],
-                defocus_ang=params["defocus_angle"],
-                Cs=params["cs"],
-                alpha=params["amplitude_contrast"],
+                pixel_size=filter_params[i, 6],
```
Comment (garrettwrong) [inline src/aspire/source/coordinates.py:376]: We could make a `CTFFilter.from_filter_param` static method instead of this block, but I think would be create confusion if using integer indices as we've done now. What do you think would be best?

File: src/aspire/utils/relion_interop.py [hunk 10]
```diff
@@ -1,11 +1,92 @@
...
+
+from aspire.storage import StarFile
+
+logger = logging.getLogger(__name__)
+
+
+# The metadata_fields dictionary below specifies default data types
+# of certain key fields used in the codebase,
+# which are originally read from Relion STAR files.
+relion_metadata_fields = {
```
Comment (garrettwrong) [inline src/aspire/utils/relion_interop.py:14]: Do you for-see any reason developers or users might want this at the second import level in the future?

**Review by garrettwrong — CHANGES_REQUESTED** [review]
Hi Chris thanks for putting this together, it looks like a lot of work.

I want to be careful about reviewing this because it touches a lot of code that is sort of foundational and has been stable for a while. I may have more questions, but there are some conversations you can address while I look it over more.  Thanks.

File: src/aspire/utils/__init__.py (hunk 4 above)
Comment (chris-langfield) [reply src/aspire/utils/__init__.py]: Sadly true, and I like those names better!

File: src/aspire/utils/relion_interop.py (hunk 10 above)
Comment (chris-langfield) [reply src/aspire/utils/relion_interop.py:14]: Added to the `__init__` thanks

File: tests/test_coordinate_source.py (hunk 2 above)
Comment (chris-langfield) [reply tests/test_coordinate_source.py]: Thanks for catching, fixed!

File: tests/test_orient_sync.py (hunk 1 above)
Comment (chris-langfield) [reply tests/test_orient_sync.py:111]: Yes, I figured wherever we're getting particles from a star file (outside of testing the 3.0 version), we should use 3.1

Comment (garrettwrong) [reply tests/test_orient_sync.py:111]: Okay, the comment at the top ot the 31.star just made me want to ask.  Thanks

File: tests/test_coordinate_source.py (hunk 3 above)
Comment (chris-langfield) [reply tests/test_coordinate_source.py:328]: Done, thanks

File: tests/test_orient_sync.py (hunk 1 above)
Comment (chris-langfield) [reply tests/test_orient_sync.py:111]: Yeah, actually the comment was there from when I thought the new file would be much more different, but it now represents the same information

On balance, I'll remove that comment to prevent confusion

File: src/aspire/source/coordinates.py (hunk 9 above)
Comment (garrettwrong) [reply src/aspire/source/coordinates.py:376]: Discussed resolving, merely a comment.

File: src/aspire/source/relion.py (hunk 5 above)
Comment (chris-langfield) [reply src/aspire/source/relion.py]: Added a test

File: src/aspire/source/coordinates.py (hunk 8 above)
Comment (chris-langfield) [reply src/aspire/source/coordinates.py:371]: This was missing before (the metadata test in `test_coordinate_source` has been updated to match).

It now matches `RelionSource`:

https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/1554bbba7f0c77b5c6756fdca8c1ec61351b1b8a/src/aspire/source/relion.py#L151-L163

File: src/aspire/utils/relion_interop.py (hunk 7 above)
Comment (chris-langfield) [reply src/aspire/utils/relion_interop.py:61]: This particular code doesn't interact with NumPy (or change the way metadata interacts with NumPy). It's moved from this `RelionSource` line:

https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/1554bbba7f0c77b5c6756fdca8c1ec61351b1b8a/src/aspire/source/relion.py#L200-L204

I'm pretty sure that the `str` default was put intentionally to cast unknown relion columns to string by default. Relion does have a lot of columns whose data will be paths to other files, etc, so it makes sense to me.

The only change putting it here makes is that every Relion starfile is just dtyped right away before being processed (e.g. by `apply_optics_group()`)

Comment (chris-langfield) [reply src/aspire/utils/relion_interop.py:61]: I don't think it interacts with the workaround you mentioned, as that is done after the metadata has already been written and saved (well beyond the STAR file parsing phase). That code is here:

https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/1554bbba7f0c77b5c6756fdca8c1ec61351b1b8a/src/aspire/source/image.py#L281-L288

File: src/aspire/source/coordinates.py (hunk 8 above)
Comment (garrettwrong) [reply src/aspire/source/coordinates.py:371]: Cool.

File: tests/saved_test_data/sample_particles_relion31.star [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+loop_ 
+_rlnOpticsGroupName #1 
+_rlnOpticsGroup #2 
+_rlnMtfFileName #3 [Modulation Transfer Function]
+_rlnMicrographOriginalPixelSize #4 
+_rlnVoltage #5 
+_rlnSphericalAberration #6 
+_rlnAmplitudeContrast #7 
+_rlnImagePixelSize #8 
+opticsGroup1    1  mtf_k2_200kV.star    0.885000    300.000000  2.700000    0.100000    1.400000
```
Comment (janden) [inline tests/saved_test_data/sample_particles_relion31.star:12]: Add a second optics group to see that we can get the values properly from both?

File: src/aspire/storage/starfile.py [hunk 12]
```diff
@@ -205,3 +205,43 @@ def __eq__(self, other):
...
+            column in data_block_columns
+            for column in ["_rlnImageName", "_rlnMicrographName", "_rlnMovieName"]
+        ):
+            return None
+
+        return "3.0"
+
+    # 3.1
+    if len(star) == 2:
+        # first block must be called "optics"
```
Comment (janden) [inline src/aspire/storage/starfile.py]: I know it's typically first, but that's not a hard requirement, is it? Besides, it looks like the code below does not check for the block index.

File: src/aspire/storage/starfile.py [hunk 13]
```diff
@@ -205,3 +205,43 @@ def __eq__(self, other):
...
+    Determines whether a supplied STAR file is of Relion 3.0 or 3.1 and on
+        format.
+    :param filepath: Path to a STAR file.
+    :return: Either "3.0" or "3.1", or `None` if ASPIRE cannot interpret this file as a RELION
+        STAR file.
+    """
+    star = StarFile(filepath)
+    # 3.0 files have one block, >=3.1 files have two
+    if len(star) == 1:
+        data_block_columns = star.get_block_by_index(0).columns.to_list()
```
Comment (janden) [inline src/aspire/storage/starfile.py]: Wouldn't it be good to check if this block has the proper name, too? It should be `"particles"`, `"images"`, etc. according to the Relion conventions doc. I think in some older files, it doesn't have a name (so it's just empty, given by a `data_` tag). All the same, it might be good to check this, save the name, and then have `data_block` or some property map to that (that is, in `Relion31StarFile`).

File: src/aspire/storage/starfile.py [hunk 14]
```diff
@@ -205,3 +205,43 @@ def __eq__(self, other):
...
+    # 3.1
+    if len(star) == 2:
+        # first block must be called "optics"
+        if "optics" not in star.blocks.keys():
+            return None
+        # optics block must contain group number
+        if "_rlnOpticsGroup" not in star["optics"].columns.to_list():
+            return None
+        # lastly, the second  block must have a column determining the type of data
+        data_block_columns = star.get_block_by_index(1).columns.to_list()
```
Comment (janden) [inline src/aspire/storage/starfile.py]: Again, might be good to index this by name (`"particles"`).

File: src/aspire/storage/starfile.py [hunk 15]
```diff
@@ -205,3 +205,43 @@ def __eq__(self, other):
                 if not self_list[i][1] == other_list[i][1]:
                     return False
         return True
+
+
+def getRelionStarFileVersion(filepath):
```
Comment (janden) [inline src/aspire/storage/starfile.py]: Snake case.

File: src/aspire/utils/relion_interop.py [hunk 16]
```diff
@@ -1,11 +1,91 @@
...
+def df_to_relion_types(df):
+    # convert STAR file strings to data type for each field
+    # columns without a specified data type are read as dtype=object
+    column_types = {name: relion_metadata_fields.get(name, str) for name in df.columns}
+    return df.astype(column_types)
+
+
+class Relion30StarFile(StarFile):
+    def __init__(self, filepath):
+        super().__init__(filepath, blocks=None)
```
Comment (janden) [inline src/aspire/utils/relion_interop.py:71]: So we don't provide the `blocks` kwarg for the subclasses. Is there a point to having it or should we remove it in `StarFile` as well? If so, maybe let's just do that now (or make an issue).

File: src/aspire/utils/relion_interop.py [hunk 17]
```diff
@@ -1,11 +1,91 @@
...
+        _blocks = OrderedDict()
+        for block_name, block in self.blocks.items():
+            _blocks[block_name] = df_to_relion_types(block)
+        self.blocks = _blocks
+
+
+class Relion31StarFile(Relion30StarFile):
+    def __init__(self, filepath):
+        super().__init__(filepath)
+        self.optics_block = self.get_block_by_index(0)
```
Comment (janden) [inline src/aspire/utils/relion_interop.py]: Safer to index by name?

File: src/aspire/utils/relion_interop.py [hunk 18]
```diff
@@ -1,11 +1,91 @@
...
-        self.cs = float(df_row._rlnSphericalAberration)
-        self.amplitude_contrast = float(df_row._rlnAmplitudeContrast)
-        self.pixel_size = float(df_row._rlnMicrographPixelSize)
+        data_block = self.data_block.copy()
+        # get a NumPy array of optics indices for each row of data
+        optics_indices = self.data_block["_rlnOpticsGroup"].astype(int).to_numpy()
+        for optics_index, row in self.optics_block.iterrows():
+            # find row indices with this optics index
+            # Note optics group number is 1-indexed in Relion
+            match = np.nonzero(optics_indices == optics_index + 1)[0]  # returns 1-tuple
```
Comment (janden) [inline src/aspire/utils/relion_interop.py]: I think `np.argmax` is preferred here (avoids the tuple thing and returns at first hit).

File: src/aspire/utils/relion_interop.py [hunk 19]
```diff
@@ -1,11 +1,91 @@
...
+        self.blocks = _blocks
+
+
+class Relion31StarFile(Relion30StarFile):
+    def __init__(self, filepath):
+        super().__init__(filepath)
+        self.optics_block = self.get_block_by_index(0)
+        self.data_block = self.get_block_by_index(1)
+
+    def apply_optics_block(self):
```
Comment (janden) [inline src/aspire/utils/relion_interop.py]: So “apply” implies some change to the object, which is not the case here. Maybe something like `merged_data_block` would be more appropriate?

File: src/aspire/source/coordinates.py [hunk 20]
```diff
@@ -273,170 +300,119 @@ def _get_mrc_shapes(self):
...
+            `CoordinateSource`.
         """
-        if not len(ctf_files) == len(self.mrc_paths):
+        relion_version = getRelionStarFileVersion(ctf)
+        if not relion_version:
             raise ValueError(
-                "Number of CTF STAR files must match number of micrographs."
+                f"Cannot recognize {ctf} as a valid Relion STAR file containing micrograph information."
+            )
+        if relion_version == "3.0":
```
Comment (janden) [inline src/aspire/source/coordinates.py]: This pattern happens twice in the PR, which makes me think that it might make sense to group them together in a simple wrapper class, `RelionStarFile` that does the version detection and has some method `get_data_block`, which does what we want.

File: src/aspire/source/coordinates.py [hunk 21]
```diff
@@ -273,170 +300,119 @@ def _get_mrc_shapes(self):
...
-            "defocus_v": float(df["_rlnDefocusV"][0]),
-            "defocus_angle": float(df["_rlnDefocusAngle"][0]),
-            "cs": float(df["_rlnSphericalAberration"][0]),
-            "amplitude_contrast": float(df["_rlnAmplitudeContrast"][0]),
-            "voltage": float(df["_rlnVoltage"][0]),
-            "pixel_size": float(df["_rlnDetectorPixelSize"][0]),
-        }
+        # set metadata
+        for mrc_idx, filter_index in enumerate(indices):
+            particle_indices_this_mrc = [
```
Comment (janden) [inline src/aspire/source/coordinates.py]: `mrc_index_to_particles[mrc_idx]`?

File: tests/test_coordinate_source.py [hunk 22]
```diff
@@ -266,7 +267,7 @@ def createTestCtfFiles(self, index):
             "_rlnSphericalAberration": 700 + index,
             "_rlnAmplitudeContrast": 600 + index,
             "_rlnVoltage": 500 + index,
-            "_rlnDetectorPixelSize": 400 + index,
+            "_rlnMicrographPixelSize": 400 + index,
```
Comment (janden) [inline tests/test_coordinate_source.py:270]: Just curious: why this change?

**Review by janden — CHANGES_REQUESTED** [review]
Looks good! Just have a couple of questions and suggestions. Nothing major.

File: src/aspire/utils/relion_interop.py [hunk 23]
```diff
@@ -1,11 +1,91 @@
...
         """
-        Creates a simple object to store optics group data given a pandas dataframe row.
+        Applies the parameters in the optics block as new columns in the data block,
+            based on the corresponding optics group number. Returns a new DataFrame.
+        :return: A new DataFrame with the optics parameters added as columns.
```
Comment (j-c-c) [inline src/aspire/utils/relion_interop.py]: blank line before `:return: ...`

File: src/aspire/storage/starfile.py [hunk 24]
```diff
@@ -205,3 +205,43 @@ def __eq__(self, other):
...
+    """
+    Determines whether a supplied STAR file is of Relion 3.0 or 3.1 and on
+        format.
+    :param filepath: Path to a STAR file.
```
Comment (j-c-c) [inline src/aspire/storage/starfile.py]: blank line before `:param ...:`

File: src/aspire/utils/relion_interop.py [hunk 25]
```diff
@@ -1,11 +1,91 @@
...
+    "_rlnNrOfFrames": int,
+    "_rlnMaxValueProbDistribution": float,
+    "_rlnOpticsGroup": int,
+    "_rlnOpticsGroupName": str,
+}
+
+
+def df_to_relion_types(df):
+    # convert STAR file strings to data type for each field
+    # columns without a specified data type are read as dtype=object
```
Comment (j-c-c) [inline src/aspire/utils/relion_interop.py]: This looks like a single comment over two lines. Maybe add periods or CAPS or both.

**Review by j-c-c — CHANGES_REQUESTED** [review]
Looks good Chris! Just some dosctring fixes.

File: src/aspire/utils/relion_interop.py (hunk 25 above)
Comment (chris-langfield) [reply src/aspire/utils/relion_interop.py]: added both

File: src/aspire/storage/starfile.py (hunk 24 above)
Comment (chris-langfield) [reply src/aspire/storage/starfile.py]: done

File: src/aspire/utils/relion_interop.py (hunk 23 above)
Comment (chris-langfield) [reply src/aspire/utils/relion_interop.py]: done thanks

File: tests/test_coordinate_source.py (hunk 22 above)
Comment (chris-langfield) [reply tests/test_coordinate_source.py:270]: We had the wrong relion metadata column here, I believe. 

The output of our CTF estimator was changed by #817 

Discussion in the CTF validation issue:

https://github.com/ComputationalCryoEM/ASPIRE-Python/issues/742#issuecomment-1361697646

**Review by j-c-c — DISMISSED** [review]
LGTM!

File: tests/test_coordinate_source.py (hunk 22 above)
Comment (janden) [reply tests/test_coordinate_source.py:270]: Ok got it.

File: src/aspire/utils/relion_interop.py (hunk 16 above)
Comment (chris-langfield) [reply src/aspire/utils/relion_interop.py:71]: We do use that for saving `Source`s:

https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/1554bbba7f0c77b5c6756fdca8c1ec61351b1b8a/src/aspire/source/image.py#L702-L709

Perhaps the `Relion3*StarFile`s should be fleshed out so we save Source's with them too? Right now it's sort of just one directional for reading the STAR files.

File: src/aspire/utils/relion_interop.py (hunk 18 above)
Comment (chris-langfield) [reply src/aspire/utils/relion_interop.py]: We need *all* the indices from the data block with the given optics index though.. 

For instance if the first 7 particles are in optics group 1, `nonzero` will give us `[0,1,2,3,4,5,6]` but `argmax` will give us just `0`.

Following your train of thought I looked into using `np.argwhere`, but the docs page actually says to use `nonzero` if indexing arrays:

https://numpy.org/doc/stable/reference/generated/numpy.argwhere.html

File: src/aspire/source/coordinates.py (hunk 21 above)
Comment (chris-langfield) [reply src/aspire/source/coordinates.py]: Thanks for catching that! Fixed

File: src/aspire/utils/relion_interop.py (hunk 19 above)
Comment (chris-langfield) [reply src/aspire/utils/relion_interop.py]: Agreed and renamed

File: src/aspire/storage/starfile.py (hunk 15 above)
Comment (chris-langfield) [reply src/aspire/storage/starfile.py]: Fixed

File: tests/saved_test_data/sample_particles_relion31.star (hunk 11 above)
Comment (chris-langfield) [reply tests/saved_test_data/sample_particles_relion31.star:12]: Done!

File: src/aspire/storage/starfile.py (hunk 12 above)
Comment (chris-langfield) [reply src/aspire/storage/starfile.py]: You're right, added some code to get the blocks by name instead

File: src/aspire/storage/starfile.py (hunk 13 above)
Comment (chris-langfield) [reply src/aspire/storage/starfile.py:222]: Done, see comment above, thanks!

File: src/aspire/storage/starfile.py (hunk 14 above)
Comment (chris-langfield) [reply src/aspire/storage/starfile.py]: Done, see comment above, thanks!

File: src/aspire/storage/starfile.py (hunk 13 above)
Comment (chris-langfield) [reply src/aspire/storage/starfile.py]: Yeah, generally it doesn't have the name in the older files.. I could implement saving the `data_block` name for the 3.1 Starfiles?

File: src/aspire/utils/relion_interop.py (hunk 16 above)
Comment (janden) [reply src/aspire/utils/relion_interop.py:71]: Ok that makes sense. Thanks! I think it's fine to leave as is, then.

File: src/aspire/utils/relion_interop.py (hunk 18 above)
Comment (janden) [reply src/aspire/utils/relion_interop.py]: Ok got it. So the 1-tuple is just because `ndims == 1` then.

File: src/aspire/storage/starfile.py (hunk 13 above)
Comment (janden) [reply src/aspire/storage/starfile.py]: Yeah that would be good. I think the name is going to depend on the type of STAR file we're dealing with. If it's indexing micrographs, the block will probably be called `"images"`, if it's indexing single particle projections, it'll be `"particles"`, and so on.

Comment (chris-langfield) [reply src/aspire/storage/starfile.py]: Going off the Relion 3.1 tutorial data,

in the Import job, it has movies in a block called data `movies`

in the CTFfind/other micrograph jobs, it has micrographs in a block called `micrographs`

And for the rest of the pipeline, generally `particles`.

Comment (chris-langfield) [reply src/aspire/storage/starfile.py]: Also, sorry now I see you specified `Relion31StarFile` in your original comment :) working on it

**Comment by chris-langfield** [comment]
Something to note for dealing with arbitrary STAR files (via `aspire.storage.starfile`) is something I had forgotten, which is that for information with one row and several named columns, `gemmi` will store the block as a dictionary, rather than a DataFrame. Came across this while addressing the review (see `RelionStarFile.covert_dtypes()`). Not sure if issue-worthy but I thought I'd mention it.

File: src/aspire/source/coordinates.py (hunk 20 above)
Comment (chris-langfield) [reply src/aspire/source/coordinates.py]: Done, let me know what you think. Thanks!

File: src/aspire/storage/starfile.py (hunk 13 above)
Comment (chris-langfield) [reply src/aspire/storage/starfile.py]: I had one additional question for you regarding this line. Given that the legacy STAR files usually have a block with blank title, do you think we'd want to access via `blocks[""]` or get the sole block by index, e.g. `star.get_block_by_index(0)` as I have above?

File: src/aspire/utils/relion_interop.py (hunk 17 above)
Comment (chris-langfield) [reply src/aspire/utils/relion_interop.py]: Addressed as part of `RelionStarFile` refactor

File: src/aspire/storage/starfile.py (hunk 13 above)
Comment (janden) [reply src/aspire/storage/starfile.py]: For the legacy STAR files, I think this is less of an issue, since we know that there's only a single block. There it may make more sense to take the only block instead of worrying about names.

File: src/aspire/utils/relion_interop.py [hunk 26]
```diff
@@ -50,42 +50,130 @@
...
 
-
-class Relion31StarFile(Relion30StarFile):
-    def __init__(self, filepath):
-        super().__init__(filepath)
-        self.optics_block = self.get_block_by_index(0)
-        self.data_block = self.get_block_by_index(1)
-
-    def apply_optics_block(self):
+    def get_data_block(self):
```
Comment (janden) [inline src/aspire/utils/relion_interop.py]: Would probably rename this `get_merged_data_block`. Otherwise it's not clear how this differs from the `data_block` attribute.

File: src/aspire/utils/relion_interop.py [hunk 27]
```diff
@@ -50,42 +50,130 @@
...
         super().__init__(filepath, blocks=None)
 
-        # first convert types where possible
+        # convert dtypes in star file blocks where possible
+        self.convert_dtypes()
+
+        # validate star file and detect Relion version (3.0 or >=3.1)
+        self.validate_and_detect_version()
+
+    def validate_and_detect_version(self):
```
Comment (janden) [inline src/aspire/utils/relion_interop.py]: This and `convert_dtypes` are only ever called internally, so maybe we should make them private?

**Review by janden — CHANGES_REQUESTED** [review]
Looks great. Just two more things.

File: src/aspire/source/coordinates.py (hunk 20 above)
Comment (janden) [reply src/aspire/source/coordinates.py]: Looks good!

File: src/aspire/utils/relion_interop.py (hunk 26 above)
Comment (chris-langfield) [reply src/aspire/utils/relion_interop.py]: Renamed

File: src/aspire/utils/relion_interop.py (hunk 27 above)
Comment (chris-langfield) [reply src/aspire/utils/relion_interop.py]: Done

**Review by janden — APPROVED** [review]

**Review by garrettwrong — APPROVED** [review]
